utils/zerovec/src/zerovec/slice.rs

```diff
@@ -0,0 +1,318 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::*;
+use alloc::boxed::Box;
+
+/// A zero-copy "slice", i.e. the zero-copy version of `[T]`. This behaves
+/// similarly to [`ZeroVec<T>`], however [`ZeroVec<T>`] is allowed to contain
+/// owned data and as such is ideal for deserialization since most human readable
+/// serialization formats cannot unconditionally deserialize zero-copy.
+///
+/// This type can be used inside `VarZeroVec<T>`: This essentially allows for
+/// the construction of zero-copy types isomorphic to `Vec<Vec<T>>` by instead
+/// using `VarZeroVec<ZeroSlice<T>>`. See the [`VarZeroVec`](crate::VarZeroVec) docs for an example.
+#[repr(transparent)]
+pub struct ZeroSlice<T: AsULE>([T::ULE]);
+
+impl<T> ZeroSlice<T>
+where
+    T: AsULE,
+{
+    /// Get this [`ZeroSlice`] as a borrowed [`ZeroVec`]
+    ///
+    /// [`ZeroSlice`] does not have most of the methods that [`ZeroVec`] does,
+    /// so it is recommended to convert it to a [`ZeroVec`] before doing anything.
+    #[inline]
+    pub fn as_zerovec(&self) -> ZeroVec<'_, T> {
+        ZeroVec::Borrowed(&self.0)
+    }
+
+    /// Construct a `&ZeroSlice<T>` from a slice of ULEs
+    #[inline]
+    pub fn from_slice(slice: &[T::ULE]) -> &Self {
```

Comment:
Thought: we are inconsistent about whether the word "slice" refers to a slice of T or a slice of T::ULE.  We also now have ZeroSlice which also uses the word "slice".  Is this okay, or should we use different terms throughout the API for each case?

Probably good to file a follow-up issue, or perhaps reuse #1181

Reply:
Yeah we should call these `ule_slice`